### ISSUE TITLE
advice

### ISSUE THREAD

**Issue by jl777** [gklingler/sharedLibsDemo#1]
I just found your repo about the OS portable shared libraries and it looks great.

I have a project where not only does it have to be OS portable, the main codebase has a architecture where there is the main control code, which then given a specific variant would need to call that variant's state machine.

At the source level, it would be a matter of having a data structure with the function pointers for each of the state machine functions. However, I cannot know all the possible variants upfront and I would like it to be forward compatible with statemachine plugins.

Not sure how best to do this...
I was thinking that the baseline code could be calling a shared library that somehow is configured to "bind" to a specific statemachine plugin. then the adding of the new statemachine can be done at the system level.

Alternatively, each statemachine could be a separate executable that registers with the baseline codebase using IPC, but this runs into runtime OS compatibility issues.

Any advice is much appreciated!

To summarize, we have an OS portable codebase (in C) that is calling a standardized statemachine through a predefined set of interface functions. There will be many different statemachine variants, and the exact type (name) of all the future statemachines is unknown.

Some sane way to create the set of OS portable statemachines that all are called by the OS portable baseline code is needed.

**Comment by gklingler** [comment]
I only know very little about the requirements, but can't the "main control code" not just dynamically load the appropriate state-machine shared lib depending on some configuration?
To be OS portable you would of course need some kind of "abstraction layer" doing the dynamic/runtime loading of the of lib(s).
Not sure if this is applicable in your project, but with the limited info this is the most straight forward soution coming to my mind.

**Comment by jl777** [comment]
I think what I need is a way to pass in a list of dynamically linked libs to the main control code. Then at init time it would iterate through this list and update the list of active statemachine names, one of the lib functions would be something that returned function pointers for the various state functions. 

Then during runtime based on the statemachine name, the right lib's function pointers would be called.

So, the question is if it is possible (sane) to have OS portable function pointer interface in dynamic libs

**Comment by gklingler** [comment]
I mean you should not dynamically link all available statemachine libs but dynamically load (dlopen/dlsym on Linux, LoadLibrary() and GetProcAddress() on Win etc.).

**Comment by jl777** [comment]
yes, that is what i tried to say, a runtime "linking", I guess loading. There is a way to encapsulate all the OS specific aspects of this?

(portable main control code) <- OS independent dynamic loading module -> OS specific dynamic loader <-> portable statemachine

so the main control code writes to the same interface for all OS, each OS gets a layer that implements that and the statemachine's are portable code talking to the other side
